temporalio/worker/workflow_sandbox/importer.py

```diff
@@ -0,0 +1,287 @@
+"""Restrictive importer for workflow sandbox.
+
+.. warning::
+    This API for this module is considered unstable and may change in future.
+"""
+
+from __future__ import annotations
+
+import builtins
+import functools
+import importlib
+import logging
+import sys
+import types
+from contextlib import contextmanager
+from typing import Any, Dict, Iterator, Mapping, Optional, Sequence, Set, Tuple
+
+from typing_extensions import Protocol
+
+import temporalio.workflow
+
+from .restrictions import (
+    RestrictedModule,
+    RestrictedWorkflowAccessError,
+    RestrictionContext,
+    SandboxRestrictions,
+)
+
+logger = logging.getLogger(__name__)
+
+# Set to true to log lots of sandbox details
+LOG_TRACE = False
+_trace_depth = 0
+
+
+def _trace(message: object, *args: object) -> None:
+    if LOG_TRACE:
+        global _trace_depth
+        logger.debug(("  " * _trace_depth) + str(message), *args)
+
+
+class Importer:
+    """Importer that restricts modules."""
+
+    def __init__(self, env: _Environment) -> None:
+        """Create importer."""
+        self.env = env
+        self.orig_modules = sys.modules
+        self.new_modules: Dict[str, types.ModuleType] = {}
+        self.orig_import = builtins.__import__
+        self.modules_checked_for_restrictions: Set[str] = set()
+        self.import_func = self._import if not LOG_TRACE else self._traced_import
+
+    # TODO(cretz): Document that this is expected to have a global lock
+    @contextmanager
+    def applied(self) -> Iterator[None]:
+        """Context manager to apply this restrictive import.
+
+        .. warning::
+            This currently alters global sys.modules and builtins.__import__
+            while it is running and therefore should be locked against other
+            code running at the same time.
+        """
+        # Error if already applied
+        if sys.modules is self.new_modules:
+            raise RuntimeError("Sandbox importer already applied")
+
+        # Set our modules, then unset on complete
+        sys.modules = self.new_modules
+        builtins.__import__ = self.import_func
```

Comment:
Can you use something like `conextvars` to ensure the replacement happens only in this context and not in other concurrently running threads? It should allow you to remove the global lock.

```python
class ContextifiedModules:
    def __getattr__(self, attr):
        return getattr(ctx_modules.get(), attr)

    def __getitem__(self, item):
        return ctx_modules.get()[item]


ctx_modules = ContextVar('modules', default=sys.modules)
sys.modules = ContextifiedModules()  # TODO: fix types
ctx_modules.set({})
```

You can do the same with the importer, it's even simpler since it's a function.

Alternatively you can proxy `getattr` for `self.new_modules["sys"]`.

Reply:
No, context vars don't cross many things by default. I could use thread local maybe. I will try that.

Also, I am not sure you can set sys modules to anything not a dict due to C-level checks. I will also try that.

Reply:
My first attempt at this has failed. The error at https://github.com/python/cpython/blob/v3.10.8/Lib/dataclasses.py#L546 is being thrown by the `dataclasses` library if I don't put the _original dict_ back on `sys.modules` `unapplied` instead of just swapping out the underlying dict.

Specifically the error is:

> TypeError: non-default argument 'all' follows default argument

This is breaking on my `SandboxMatcher` dataclass because I have fields like:

```
    only_runtime: bool = False

    all: ClassVar[SandboxMatcher]
```

This is related to https://bugs.python.org/issue33453 but of course doesn't happen with the regular sys.modules, only my replaced one. Putting `sys.modules` back on `unapplied` fixes it. I think this has to do with the use of globals inside https://github.com/python/cpython/blob/v3.10.8/Lib/dataclasses.py#L882.

This is the type of thing that will take a day do debug (these come up all the time on this issue, hence why I say I had to do fickle global state replacement)

Reply:
Ok, this was a bug in my thread-local MutableMapping impl. Continuing on...

Reply:
Fixed this by using thread-local for my global pieces. It is a great improvement and solved one of my biggest concerns about the impl.